### PR TITLE
Feat/authentication/form-validation : Integrate form validation library into authentication screens

### PR DESCRIPTION
**Pull request by charliemangano**
# Integrate form validation library into authentication screens

## Description
This PR integrate a new library [FromBuilder](https://github.com/jkuatdsc/form-builder) to do form validation in our authentication screens. This library is made to work with Jetpack Compose. This tremendously cleans up our screens since the validation logic is handled elsewhere, and better separates the duties MVVM-wise.
This PR closes issue #262 and part of #261.

## Changes
**Explanation of the library:**
FormBuilder (docs [here](https://jkuatdsc.github.io/form-builder/form-builder/com.dsc.form_builder/index.html)) provides us with a `FormState`, which is a `State` to represent a form. Here, we only use fields, but it also provides support for checkboxes, radio buttons, etc. 

The `FormState` has a list of `TextFieldState`s passed as arguments on initialization, to represent each one of the fields (here, email, password and password confirmation). 

Each `TextFieldState` has an identifying name and a list of `Validator`s, that provide "rules" that the field input has to follow to be considered valid. 

These `Validator`s are either provided by the library (like `Required` or `Min` for example) or custom (built using the `Custom` validator). Each of them have an error message.

Using the `getState` function, we can extract the `TextFieldState` by name. Using the `change` function, we can trigger a recompose with the new value. Using `validate`, we can evaluate whether or not the value of the input is valid (follows the rules defined by the associated `Validator`s) or not. We can access the error message with `errorMessage`.

**Actual changes:**
I set up the validation logic in the `AuthenticationViewModel`: the `FormState` is accessible as an attribute of the class and is associated with the corresponding `TextFieldState`s and `Validator`s.
In the screens, I replaced the existing states with the `formState` and offshoots `emailState`, `passwordState` etc. I also replaced the validation logic with calls to `validate` and the `onChange` logic with calls to `change`. The error messages are also better handled using the `errorMessage` function.

Seeing the feedback from M2, I added a check for the length of the inputs. FormBuilder also implements proper checks for the email so just having a `@` is not sufficient anymore. The email needs to be of the form `<smth>@<smth>.<smth>`.

#### Added
None
#### Modified
- `app/src/main/java/com/android/periodpals/model/authentication/AuthenticationViewModel.kt`: implemented the validation logic
- `app/src/main/java/com/android/periodpals/ui/authentication/SignIn.kt` and `app/src/main/java/com/android/periodpals/ui/authentication/SignUp.kt`: integrated the library into the screens
- `app/src/test/java/com/android/periodpals/model/authentication/AuthenticationViewModelTest.kt`, `app/src/test/java/com/android/periodpals/ui/authentication/SignInTest.kt` and `app/src/test/java/com/android/periodpals/ui/authentication/SignUpTest.kt`: update tests
- `app/build.gradle.kts`, `settings.gradle.kts` and `gradle/libs.versions.toml`: added FormBuilder dependency

#### Removed
None

## Dependencies Added
- `com.github.jkuatdsc:form-builder`: FormBuilder dependency

## Testing
I updated the tests to take into account this new library: in `AuthenticationViewModelTest`, I implemented new checks that verify the `FormState` attribute values, and in `SignInTest` and `SignUpTest`, I provided the `FormState` as part of the `AuthenticationViewModel` mock.

### CODE REVIEW

**Review by francelu — CHANGES_REQUESTED** [review]
## Summary
This PR is a good clean up to validate fields. As the coaches mentioned, we're now using this Form Builder API. I found it really interesting to see how it worked. Good job doing these implementations !  

## Important
#### Code Quality
- Code is clean simple and consistent with rest of the code.
- I noticed you often do some code formatting clean-up using the `,`. Good, but I think not that meaningful. Of course it's nice visually and more consistent. However, I feel like it's a loss of time as we really want to focus on implementing more features ... xD

#### Functionality
- At first sight, the implementation looks good. While in reality, nothing works, not even login (on emulator).

#### Performance
- No significant performance concerns detected.
- Likely to improve performance.

#### Security
- No specific vulnerabilities noted.

## Testing
- Tests are updated to reflect the new validation logic and the use of `FormState`.  
- It’s great to see comprehensive tests for both the ViewModel and UI. Ensure edge cases like rapid user input or empty fields are covered.
- Some cases shouldn't happen. See screenshots.

## Documentation
- The PR description is thorough and provides a clear explanation of the `FormBuilder` library and how it has been integrated. Excellent job here!
- Missing a couple of updates on existing documentation (see comments).

## Suggestions
- Check comments for nitpick, documentation, suggestions / for later task.
- Check the screenshots bellow, nothing works :((
- Consider avoiding over-formatting unless it is strictly necessary. It’s nice for consistency but can be time-consuming without added functionality.  

## Steps before PR approved
- Check code comments.
- Check the screen shots below.
- Fix the authentication.
- Might need to change the tests or add some tests because they all pass while authentication doesn't work a bit in reality.

## Screenshots
| Password error message missing on `SignInScreen`  | image  |
|--|--|
| One of my valid email addresses with wrong password doesn't display error messages correctly  | image  |
| Can't login whaaaat  | image  |
| Error messages of don't match are not displayed  | image  |

**Comment by charliemangano** [comment]
Hi @francelu, I'll respond to some things here!

> I noticed you often do some code formatting clean-up using the ,. Good, but I think not that meaningful. Of course it's nice visually and more consistent. However, I feel like it's a loss of time as we really want to focus on implementing more features ... xD

This is just a feature of my ktfmt settings, it adds it on its own when I run it!


> One of my valid email addresses with wrong password doesn't display error messages correctly

That's normal, as the checks for a valid email address became more strict: you need to have the format `<smth>@<smth>.<smth>`. Here, you're missing the `.<smth>`!


> Can't login whaaaat

You can't login because the email address is not valid. Even if you have an associated account in Supabase, this check catches the invalid email first and the login request never even goes to Supabase.


> Error messages of don't match are not displayed

This is normal as the email address is checked for first. If you try with a valid email address, the "Passwords do not match" error will show up.


> Password error message missing on SignInScreen

I understand this is the exact same behavior as before we integrated this new library, but it's still proper UX. I will try and make it consistent with how it was before, but since the tests still passed I thought it was not expected behavior.

File: app/src/main/java/com/android/periodpals/ui/authentication/SignIn.kt [hunk 1]
```diff
@@ -131,27 +121,25 @@ fun SignInScreen(
...
         AuthenticationPasswordInput(
-            password = password,
-            onPasswordChange = { password = it },
+            password = passwordState.value,
+            onPasswordChange = { passwordState.change(it) },
             passwordVisible = passwordVisible,
             onPasswordVisibilityChange = { passwordVisible = !passwordVisible },
-            passwordErrorMessage = passwordErrorMessage,
+            passwordErrorMessage = passwordState.errorMessage,
         )
```
[the remark this answers is not in the document]
Comment (charliemangano) [reply app/src/main/java/com/android/periodpals/ui/authentication/SignIn.kt:135]: This is out of the scope of my issue, but it's a very good idea. I created a task "Improve UX for input fields".

File: app/src/main/java/com/android/periodpals/ui/authentication/SignUp.kt [hunk 2]
```diff
@@ -37,32 +37,15 @@ import com.android.periodpals.ui.components.GradedBackground
...
+import com.dsc.form_builder.TextFieldState
 
-private const val DEFAULT_EMAIL = ""
-private const val DEFAULT_PASSWORD = ""
-private const val DEFAULT_CONFIRMED_PASSWORD = ""
-private const val DEFAULT_EMAIL_INVALID_MESSAGE = ""
-private const val DEFAULT_PASSWORD_INVALID_MESSAGE = ""
-private const val DEFAULT_CONFIRMED_PASSWORD_INVALID_MESSAGE = ""
 private const val DEFAULT_PASSWORD_VISIBLE = false
 private const val DEFAULT_CONFIRMED_PASSWORD_VISIBLE = false
```
[the remark this answers is not in the document]
Comment (charliemangano) [reply app/src/main/java/com/android/periodpals/ui/authentication/SignUp.kt:43]: I'll make the naming consistent between both screens.
As for the use of a `private const val` for storing this value, I still think it's good practice. I'll explicit the type in the declaration and adjust the naming to something that makes you think more of a `Boolean`. I think since they're conceptually the same, I can also make it so that there is only a single `val` for this.

File: app/src/main/java/com/android/periodpals/ui/authentication/SignUp.kt [hunk 3]
```diff
@@ -197,29 +172,27 @@ fun SignUpScreen(
...
-    setEmailErrorMessage: (String) -> Unit,
-    setPasswordErrorMessage: (String) -> Unit,
-    setConfirmedPasswordErrorMessage: (String) -> Unit,
+    emailState: TextFieldState,
+    passwordState: TextFieldState,
+    confirmPasswordState: TextFieldState,
     authenticationViewModel: AuthenticationViewModel,
     context: Context,
     navigationActions: NavigationActions,
 ) {
```
[the remark this answers is not in the document]
Comment (charliemangano) [reply app/src/main/java/com/android/periodpals/ui/authentication/SignUp.kt:181]: Will adjust! Nice catch

File: app/src/test/java/com/android/periodpals/ui/authentication/SignInTest.kt [hunk 4]
```diff
@@ -40,23 +46,52 @@ class SignInScreenTest {
...
+            ))
         .thenAnswer {
           val onSuccess = it.arguments[2] as () -> Unit
           onSuccess()
         }
     `when`(navigationActions.currentRoute()).thenReturn(Screen.SIGN_IN)
     `when`(authViewModel.userAuthenticationState)
         .thenReturn(mutableStateOf(UserAuthenticationState.Success("User is logged in")))
+    `when`(authViewModel.formState).thenReturn(formState)
+
```
[the remark this answers is not in the document]
Comment (charliemangano) [reply app/src/test/java/com/android/periodpals/ui/authentication/SignInTest.kt:94]: Since the view model is mocked in the test, we need to return a value for when `authViewModel.formState` is called. This is what this line does.

The `formState` is declared and initialised in the `setUp`, with values equal to what is in the view model. It's not mocked since it's an external library, so it's passed as its own object, and that object acts on its own as it would in the code. The `formState` is already tested in the view model tests so I don't recheck it here.

I understand the confusion, but this simply stems from the fact that `formState` is not mocked, so you don't need lines like `when(formState.someMethod).thenReturn(someValue)`.

File: app/src/main/java/com/android/periodpals/ui/authentication/SignUp.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (francelu) [reply app/src/main/java/com/android/periodpals/ui/authentication/SignUp.kt:43]: Yes a single shared val would be good enough :)

**Comment by francelu** [comment]
> This is just a feature of my ktfmt settings, it adds it on its own when I run it!

My bad, just keep doing it then ! :)

> That's normal, as the checks for a valid email address became more strict: you need to have the format <smth>@<smth>.<smth>. Here, you're missing the .<smth>!

My bad ://// I'll retry and approved if everything LGTM :))

**Review by francelu — APPROVED** [review]
## Summary
- The new changes are good, they work well and improve UX.  
- My bad for not having read thoroughly the documentation of Form Builder validation for emails.
- Everything LGTM ! :)

## Functionality
- Email is kept across the sign-in and sign-up screens. You might want to fix that, but it's a minor issue that doesn't block the user to login / sign-up.

## Documentation
- As mentioned earlier, two little updates :)

## Steps before PR merged
- Update the documentation as mentioned.
- Resolve comments.
- Then all good to go ! Good job !!

## Screenshots
| **Wrong email correct error messages**  | image  |
|--|--|
| **Keeping track of email entered en SignIn screen**  | image  |
| **Right error messages in SignUp screen**  | image  |
